proving_system/stark/src/prover.rs

```diff
@@ -4,231 +4,331 @@ use super::{
     sample_z_ood,
 };
 use crate::{
-    proof::{StarkProof, StarkQueryProof},
+    proof::{DeepConsistencyCheck, StarkProof, StarkQueryProof},
     transcript_to_field, transcript_to_usize,
 };
-use lambdaworks_crypto::fiat_shamir::transcript::Transcript;
+use lambdaworks_crypto::{
+    fiat_shamir::transcript::Transcript,
+    hash::traits::IsCryptoHash,
+    merkle_tree::{merkle::MerkleTree, proof::Proof},
+};
 use lambdaworks_math::{
     fft::errors::FFTError,
     field::{element::FieldElement, traits::IsTwoAdicField},
     polynomial::Polynomial,
     traits::ByteConversion,
 };
 
-// FIXME remove unwrap() calls and return errors
-pub fn prove<F: IsTwoAdicField, A: AIR<Field = F>>(trace: &TraceTable<F>, air: &A) -> StarkProof<F>
-where
-    FieldElement<F>: ByteConversion,
-{
-    let transcript = &mut Transcript::new();
-
-    let blowup_factor = air.options().blowup_factor as usize;
-    let coset_offset = FieldElement::<F>::from(air.options().coset_offset);
-
-    let root_order = air.context().trace_length.trailing_zeros();
-    // * Generate Coset
-    let trace_primitive_root = F::get_primitive_root_of_unity(root_order as u64).unwrap();
-    let trace_roots_of_unity = F::get_powers_of_primitive_root_coset(
-        root_order as u64,
-        air.context().trace_length,
-        &FieldElement::<F>::one(),
-    )
-    .unwrap();
-
-    let lde_root_order = (air.context().trace_length * blowup_factor).trailing_zeros();
-    let lde_roots_of_unity_coset = F::get_powers_of_primitive_root_coset(
-        lde_root_order as u64,
-        air.context().trace_length * blowup_factor,
-        &coset_offset,
-    )
-    .unwrap();
-
-    let trace_polys = trace.compute_trace_polys();
-    let lde_trace_evaluations = trace_polys
-        .iter()
-        .map(|poly| {
-            poly.evaluate_offset_fft(
-                &FieldElement::<F>::from(air.options().coset_offset),
-                air.options().blowup_factor as usize,
-            )
-        })
-        .collect::<Result<Vec<Vec<FieldElement<F>>>, FFTError>>()
+pub struct Prover<'a, F: IsTwoAdicField, A: AIR<Field = F>> {
+    air: &'a A,
+    transcript: Transcript,
+}
```

Comment:
Let's revert this `Prover` struct since having the transcript as a field can get very error prone, as we would need to make sure to reset the transcript's state if new calls to prove are made, etc.

Reply:
Done